python/cudf/cudf/dataframe/column.py

```diff
@@ -461,23 +464,94 @@ def __getitem__(self, arg):
                 # slicing data
                 subdata = self.data[arg]
                 # slicing mask
+                if self.dtype == "object":
+                    data_size = self.data.size()
+                else:
+                    data_size = self.data.size
                 bytemask = cudautils.expand_mask_bits(
-                    self.data.size, self.mask.to_gpu_array()
+                    data_size, self.mask.to_gpu_array()
                 )
                 submask = Buffer(cudautils.compact_mask_bytes(bytemask[arg]))
                 col = self.replace(data=subdata, mask=submask)
                 return col
             else:
                 newbuffer = self.data[arg]
                 return self.replace(data=newbuffer)
-        elif isinstance(arg, (list, np.ndarray)):
-            arg = np.array(arg)
-            arg = rmm.to_device(arg)
+        else:
+            arg = columnops.as_column(arg)
+            if len(arg) == 0:
+                arg = columnops.as_column([], dtype="int32")
+            if pd.api.types.is_integer_dtype(arg.dtype):
+                return self.take(arg.data.mem)
+            elif pd.api.types.is_bool_dtype(arg.dtype):
+                return self.apply_boolean_mask(arg)
+            else:
+                raise NotImplementedError(type(arg))
+
+    def __setitem__(self, key, value):
+        """
+        Set the value of self[key] to value.
+
+        If value and self are of different types,
+        value is coerced to self.dtype
+        """
+        import cudf.bindings.copying as cpp_copying
+        from cudf.dataframe import columnops
+
+        if isinstance(key, slice):
+            key_start, key_stop, key_stride = key.indices(len(self))
+            if key_stride != 1:
+                raise NotImplementedError("Stride not supported in slice")
+            nelem = abs(key_stop - key_start)
+        else:
+            key = columnops.as_column(key)
+            if pd.api.types.is_bool_dtype(key.dtype):
+                if not len(key) == len(self):
+                    raise ValueError(
+                        "Boolean mask must be of same length as column"
+                    )
+                key = columnops.as_column(cudautils.arange(len(self)))[key]
```

Comment:
I'm confused by the impl of `arange`: `size = (len - 1) // 2`?. what's happening here?

Reply:
I think operator precedence may be responsible for this, it's actually:

`size = ((len - 1) // step) + 1`

Reply:
I think I had some cognitive dissonance between `//` being floor division vs a c-style comment, and that just came out as + being height precedence. :joy:

Reply:
It appears to be [a fast implementation](https://stackoverflow.com/questions/2745074/fast-ceiling-of-an-integer-division-in-c-c/2748128#2748128) of `ceil` that doesn't require casting to float. I think the intention is to compute `ceil( (stop - start) / step)` which should come out to be the number of elements in the resulting `arange`.